### PR TITLE
feat: Ensure mRID for messages are valid in ebix

### PR DESCRIPTION
**Pull request by MWO1024**
## Description
The aim of this PR is to ensure that all EDI-generated message header ids are valid in both CIM and eBix. We achieve this, by adding a new `MessageId` to the `Bundle`. This new id is used for the header, but internally, we are still referring the `BundleId` as using proper UUIDs is preferable wrt e.g. foreign keys and the like.

New ArchivedMessages table
billede

New Bundles table
billede

MarketDocuments table
billede

OutgoingMessages table
billede

Example CIM XML output

```
<cim:NotifyAggregatedMeasureData_MarketDocument xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="urn:ediel.org:measure:notifyaggregatedmeasuredata:0:1 urn-ediel-org-measure-notifyaggregatedmeasuredata-0-1.xsd" xmlns:cim="urn:ediel.org:measure:notifyaggregatedmeasuredata:0:1">
  <cim:mRID>9a415532ab094de48d8565b28062a448</cim:mRID>
  <cim:type>E31</cim:type>
  <cim:process.processType>D04</cim:process.processType>
  <cim:businessSector.type>23</cim:businessSector.type>
  <cim:sender_MarketParticipant.mRID codingScheme="A10">5790001330552</cim:sender_MarketParticipant.mRID>
  <cim:sender_MarketParticipant.marketRole.type>DGL</cim:sender_MarketParticipant.marketRole.type>
  <cim:receiver_MarketParticipant.mRID codingScheme="A10">1111111111111</cim:receiver_MarketParticipant.mRID>
  <cim:receiver_MarketParticipant.marketRole.type>DDQ</cim:receiver_MarketParticipant.marketRole.type>
  <cim:createdDateTime>2024-07-01T14:57:09Z</cim:createdDateTime>
  <cim:Series>
    <cim:mRID>4f124e40f15d4baf9ee191b4551da0dc</cim:mRID>
    <cim:version>17198458290000000</cim:version>
    <cim:originalTransactionIDReference_Series.mRID>12356478912356478912356478912356478</cim:originalTransactionIDReference_Series.mRID>
```

Tests are green locally (will deploy to dev before merge)
billede

### CODE REVIEW

File: source/OutgoingMessages.Domain/Models/OutgoingMessages/OutgoingMessageBundle.cs [hunk 1]
```diff
@@ -61,7 +61,7 @@ public class OutgoingMessageBundle
         /// </summary>
         public Receiver DocumentReceiver { get; }
 
-        public BundleId AssignedBundleId { get; private set; }
+        public GloriousEbixUuid MessageId { get; }
```
Comment (lasrinnil) [inline source/OutgoingMessages.Domain/Models/OutgoingMessages/OutgoingMessageBundle.cs:64]: This seems a bit out of place to me - perhaps you could enlighten me why the `Bundle` no longer need a `BundleId`

File: source/BuildingBlocks.Domain/Models/TransactionId.cs [hunk 2]
```diff
@@ -46,11 +46,7 @@ public static TransactionId From(string transactionId)
 
     public static TransactionId New()
     {
-        // A normal UUID is 36 characters long, but unfortunately, the EBIX scheme only allows for 35 characters.
-        // To make everyone happy---i.e. ensure unique ids (for most practical purposes anyway) and ensure
-        // valid EBIX values---we'll just remove the dashes from the UUID.
-        return new TransactionId(
-            Guid.NewGuid().ToString().Replace("-", string.Empty, StringComparison.InvariantCultureIgnoreCase));
+        return new TransactionId(GloriousEbixUuid.NewGloriousEbixUuid().Id);
```
Comment (lasrinnil) [inline source/BuildingBlocks.Domain/Models/TransactionId.cs:49]: do we HAVE to ensure all `TransactionId`s support Ebix? What about all the CIM messages created?

File: source/BuildingBlocks.Domain/Models/GloriousEbixUuid.cs [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+namespace Energinet.DataHub.EDI.BuildingBlocks.Domain.Models;
+
+public readonly record struct GloriousEbixUuid
```
Comment (lasrinnil) [inline source/BuildingBlocks.Domain/Models/GloriousEbixUuid.cs:17]: attempt to merge with `TransactionId`

File: source/ApplyDBMigrationsApp/Scripts/Model/202405241005 Add MessageId to bundles.sql [hunk 4]
```diff
@@ -0,0 +1,13 @@
+ALTER TABLE [dbo].[Bundles]
+    ADD [MessageId] NVARCHAR(16) NULL;
```
Comment (lasrinnil) [inline source/ApplyDBMigrationsApp/Scripts/Model/202405241005 Add MessageId to bundles.sql:2]: 😢

Comment (lasrinnil) [reply source/ApplyDBMigrationsApp/Scripts/Model/202405241005 Add MessageId to bundles.sql:2]: 36

File: source/OutgoingMessages.Domain/Models/OutgoingMessages/OutgoingMessageBundle.cs (hunk 1 above)
Comment (MWO1024) [reply source/OutgoingMessages.Domain/Models/OutgoingMessages/OutgoingMessageBundle.cs]: The outgoing message bundle only need a reference to the mRID that is supposed to be present in the header element of the result.

File: source/IntegrationTests/Application/ArchivedMessages/WhenArchivedMessageIsCreatedTests.cs [hunk 5]
```diff
@@ -83,15 +83,16 @@ public async Task Archived_document_can_be_retrieved_with_correct_content()
     [InlineData(ArchivedMessageType.OutgoingMessage)]
     public async Task Archived_document_is_saved_at_correct_path(ArchivedMessageType archivedMessageType)
     {
-        var messageId = Guid.NewGuid();
+        var messageId = Guid.NewGuid().ToString();
```
Comment (lasrinnil) [inline source/IntegrationTests/Application/ArchivedMessages/WhenArchivedMessageIsCreatedTests.cs:86]:
```suggestion
        var messageId = MessageId.New();
```
?

File: source/OutgoingMessages.Interfaces/Models/PeekResultDto.cs [hunk 6]
```diff
@@ -14,7 +14,8 @@
 
 using System;
 using System.IO;
+using Energinet.DataHub.EDI.BuildingBlocks.Domain.Models;
 
 namespace Energinet.DataHub.EDI.OutgoingMessages.Interfaces.Models;
 
-public record PeekResultDto(Stream? Bundle, Guid? MessageId = default);
+public record PeekResultDto(Stream? Bundle, Guid? BundleId, MessageId? MessageId);
```
Comment (lasrinnil) [inline source/OutgoingMessages.Interfaces/Models/PeekResultDto.cs:21]: Investigate if it is possibly to remove the `BundleId`

File: source/Tests/Infrastructure/OutgoingMessages/Asserts/AssertEbixDocument.cs [hunk 7]
```diff
@@ -119,17 +112,7 @@ public async Task<AssertEbixDocument> HasValidStructureAsync(DocumentType type,
 
         if (!validationResult.IsValid && skipIdentificationLengthValidation)
```
Comment (lasrinnil) [inline source/Tests/Infrastructure/OutgoingMessages/Asserts/AssertEbixDocument.cs:113]: Time to remove this if? :)

File: source/Tests/Infrastructure/OutgoingMessages/NotifyAggregatedMeasureData/NotifyAggregatedMeasureDataDocumentWriterTests.cs [hunk 8]
```diff
@@ -283,6 +277,32 @@ await AssertDocument(document, DocumentFormat.FromName(documentFormat))
             .DocumentIsValidAsync();
     }
 
+    [Fact]
+    public async Task Can_create_ebix_document_with_36_chars_transaction_id()
```
Comment (lasrinnil) [inline source/Tests/Infrastructure/OutgoingMessages/NotifyAggregatedMeasureData/NotifyAggregatedMeasureDataDocumentWriterTests.cs:281]:
```suggestion
    public async Task support_excisting_ebix_documenst_with_36_chars_ids()
```

File: source/Tests/Infrastructure/OutgoingMessages/NotifyWholesaleServices/NotifyWholesaleServicesDocumentWriterTests.cs [hunk 9]
```diff
@@ -477,6 +471,35 @@ await AssertDocument(document, documentFormat)
             .DocumentIsValidAsync();
     }
 
+    [Fact]
+    public async Task Can_create_notifyWholesaleServices_ebix_document_with_36_chars_transaction_id()
+    {
+        // Arrange
+        var transactionId = TransactionId.From("93dbd8bb-4fbb-4b9d-b57f-f6a5c16f7bdf");
+        var originalTransactionId = TransactionId.From("b340db36-ef97-4515-839c-1d8b544e9174");
+
```
Comment (lasrinnil) [inline source/Tests/Infrastructure/OutgoingMessages/NotifyWholesaleServices/NotifyWholesaleServicesDocumentWriterTests.cs:480]: `MessageId`

File: source/Tests/Infrastructure/OutgoingMessages/NotifyWholesaleServices/NotifyWholesaleServicesDocumentWriterTests.cs [hunk 10]
```diff
@@ -477,6 +471,35 @@ await AssertDocument(document, documentFormat)
             .DocumentIsValidAsync();
     }
 
+    [Fact]
+    public async Task Can_create_notifyWholesaleServices_ebix_document_with_36_chars_transaction_id()
```
Comment (lasrinnil) [inline source/Tests/Infrastructure/OutgoingMessages/NotifyWholesaleServices/NotifyWholesaleServicesDocumentWriterTests.cs:475]: rename to make sure that the test is in place to ensure that we have combability for already created outoingmessages with ids over 35 charrs

File: source/OutgoingMessages.Domain/Models/Bundles/Bundle.cs [hunk 11]
```diff
@@ -57,10 +56,26 @@ internal Bundle(BundleId id, BusinessReason businessReason, DocumentType documen
     /// </summary>
     public MessageId? RelatedToMessageId { get; private set; }
 
+    public MessageId MessageId { get; private set; }
+
+    internal DocumentType DocumentTypeInBundle { get; }
+
+    internal BundleId Id { get; }
```
Comment (lasrinnil) [inline source/OutgoingMessages.Domain/Models/Bundles/Bundle.cs:63]: Lets give this prop a summary to make it explicit that the `BundleId` is only used internally and never exposed towards the actors.

File: source/OutgoingMessages.Interfaces/Models/PeekResultDto.cs (hunk 6 above)
Comment (MWO1024) [reply source/OutgoingMessages.Interfaces/Models/PeekResultDto.cs]: bundleId is gone from the DTO

File: source/Tests/Infrastructure/OutgoingMessages/Asserts/AssertEbixDocument.cs (hunk 7 above)
Comment (MWO1024) [reply source/Tests/Infrastructure/OutgoingMessages/Asserts/AssertEbixDocument.cs:113]: I have removed it.

**Review by lasrinnil — DISMISSED** [review]

**Review by lasrinnil — APPROVED** [review]

**Pull request merged**